content/unpack.md

```diff
@@ -0,0 +1,1084 @@
+---
+layout: sip
+permalink: /sips/:title.html
+stage: implementation
+status: under-review
+title: SIP-XX - Unpack Case Classes into Parameter Lists and Argument Lists
+---
+
+**By: Li Haoyi**
+
+## History
+
+| Date        | Version            |
+|-------------|--------------------|
+| 23 Aug 2024 | Initial Draft      |
+
+
+## Summary
+
+This proposal provides a syntax to "unpack" a `case class` _type_ into a definition-site 
+parameter list via the `unpack` keyword, and to "unpack" `case class` _value_ into a
+definition-site argument list via `*`: 
+
+```scala
+case class RequestConfig(url: String, 
+                         connectTimeout: Int,
+                         readTimeout: Int)
+
+def downloadSimple(unpack config: RequestConfig) = doSomethingWith(config)
+def downloadAsync(unpack config: RequestConfig, ec: ExecutionContext) = doSomethingWith(config)
+def downloadStream(unpack config: RequestConfig) = doSomethingWith(config)
+
+// Call with individual parameters
+val data = downloadSimple("www.example.com", 1000, 10000)
+val futureData = downloadAsync("www.example.com", 1000, 10000, ExecutionContext.global)
+val stream = downloadStream(url = "www.example.com", connectTimeout = 1000, readTimeout = 10000)
+
+// Call with config object
+val config = RequestConfig("www.example.com", 1000, 10000)
+val data2 = downloadSimple(config*)
+val futureData2 = downloadAsync(config*, ExecutionContext.global)
+val stream2 = downloadStream(config*)
+```
+
+The delegation performed by `unpack` is very similar to inheritance with
+`extends`, or composition with `export`. Scala has always had good ways to DRY up repetitive
+member definitions, but has so far had no good way to DRY up repetitive parameter lists.
+`unpack` provides the way to do so, and removes the dilemma of passing things around as loose
+parameters or `case class` values by making it easy to convert between them in both directions.
+
+## Motivation
+
+This proposal removes a tremendous amount of boilerplate converting between data structures
+and method calls in Scala. For example, the code snippet above without this feature would 
+have the parameter list duplicated many times, and any calling the methods with data from
+a `RequestConfig` object with matching fields also requires lots of duplication:
+
+```scala
+case class RequestConfig(url: String, 
+                         connectTimeout: Int,
+                         readTimeout: Int)
+
+def downloadSimple(url: String,
+                   connectTimeout: Int,
+                   readTimeout: Int) = doSomethingWith(config)
+def downloadAsync(url: String,
+                  connectTimeout: Int,
+                  readTimeout: Int,
+                  ec: ExecutionContext) = doSomethingWith(config)
+def downloadStream(url: String,
+                   connectTimeout: Int,
+                   readTimeout: Int) = doSomethingWith(config)
```

Comment:
(... continuing from above)

And here, the `doSomethingWith()` bits are downright confusing, leaving me going, "`config`?  Where does `config` come into it?"

So I'd recommend either fleshing out the example method bodies, or omitting them entirely, but this pseudo-code is causing more confusion than good.